The EUC-TW encoding doesn't have a canonical name

litherum opened the pull request:
#### 996fc9db28be5ace21aed6dfccc9290d3f87d142
<pre>
The EUC-TW encoding doesn&apos;t have a canonical name
<a href="https://bugs.webkit.org/show_bug.cgi?id=246242">https://bugs.webkit.org/show_bug.cgi?id=246242</a>
rdar://100893154

Reviewed by Alex Christensen.

ucnv_getCanonicalName(&quot;EUC-TW&quot;, &quot;IANA&quot;) is returning nullptr, which is causing
us to memcmp(nullptr, ...). This patch reorganizes the file to call ucnv_getName()
in the situations where ucnv_getCanonicalName() is returning nullptr, thereby
elliminating the need for all the special cases.

Test: imported/w3c/web-platform-tests/encoding/unsupported-labels.html

* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerCodecs):
(PAL::TextCodecICU::TextCodecICU):

Canonical link: <a href="https://commits.webkit.org/256438@main">https://commits.webkit.org/256438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df96a4ffc93e3a0bb7f3725fb9d52564e5b91240

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95733 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105309 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165632 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99716 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5066 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33746 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88113 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101149 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3726 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82349 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30780 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87489 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73609 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39478 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37173 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4451 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41239 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39596 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
```diff
@@ -115,42 +115,15 @@ void TextCodecICU::registerCodecs(TextCodecRegistrar registrar)
         ASSERT(U_SUCCESS(error));
+        if (!canonicalConverterName) {
+            auto converter = ICUConverterPtr { ucnv_open(name, &error) };
+            assert(U_SUCCESS(error));
```
Lowercase assert is a different function that we don't use in WebKit.

achristensen07 commented on Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
```diff
@@ -115,42 +115,15 @@ void TextCodecICU::registerCodecs(TextCodecRegistrar registrar)
+            canonicalConverterName = ucnv_getName(converter.get(), &error);
+            assert(U_SUCCESS(error));
+        }
         registrar(name, [name, canonicalConverterName] {
```
It probably wouldn't hurt to add another if (!canonicalConverterName) { ASSERT_NOT_REACHED(); continue; } to be on the safe side.

litherum committed:
The EUC-TW encoding doesn't have a canonical name
https://bugs.webkit.org/show_bug.cgi?id=246242
rdar://100893154

Reviewed by Alex Christensen.

ucnv_getCanonicalName("EUC-TW", "IANA") is returning nullptr, which is causing
us to memcmp(nullptr, ...). This patch reorganizes the file to call ucnv_getName()
in the situations where ucnv_getCanonicalName() is returning nullptr, thereby
elliminating the need for all the special cases.

Test: imported/w3c/web-platform-tests/encoding/unsupported-labels.html

* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerCodecs):
(PAL::TextCodecICU::TextCodecICU):

Canonical link: https://commits.webkit.org/256438@main
